design/pv-migration.md

```diff
@@ -0,0 +1,87 @@
+
+**Note:** We can not edit PV volumeAttributes with the new pool name as it is immutable field.
+
+The migrator will look for all the volumes for all the pools present on the node and will look for corresponding ZFSVolume object and will update it with the correct poolname and ownernodeId field.
+
+### Workflow
+
+- user will setup all the nodes and setup the ZFS pool on each of those nodes.
+- the ZFS-LocalPV CSI driver will look for all the pools on the node and will set the `guid.openebs.io/<pool-guid>=true` label for all ZFS POOLs that is present on that node. Let's say node-1 has two pools(say pool1 with guid as 14820954593456176137 and pool2 with guid as 16291571091328403547) present then the labels will be like this :
+```
+$ kubectl get node pawan-node-1 --show-labels
+NAME           STATUS   ROLES    AGE    VERSION   LABELS
+node-1   Ready    worker   351d   v1.17.4   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,kubernetes.io/arch=amd64,kubernetes.io/hostname=node-1,kubernetes.io/os=linux,node-role.kubernetes.io/worker=true,openebs.io/nodeid=node1,openebs.io/nodename=node-1,guid.openebs.io/14820954593456176137=true,guid.openebs.io/16291571091328403547=true
+```
+- If we are moving the pool1 from node1 to node2, then there are two cases here :-
+
+#### 1. if node2 is a fresh node
+
+- we can simply import the pool and restart the ZFS-LocalPV driver to make it aware of that pool to set the corresponding node topology
+- the ZFS-LocalPV driver will look for `guid.openebs.io/14820954593456176137=true` and will remove the label from the nodes where pool is not present
+- the ZFS-LocalPV driver will update the new node with `guid.openebs.io/14820954593456176137=true` label
+- the migrator will look for ZFSVolume resource and update the OwnerNodeID with the new node id for all the volumes.
+- the k8s scheduler will be able to see the new label and should schedule the pods to this new node.
+
+#### 2. if node2 is existing node and Pool of the same name is present there
+
+- here we need to import the pool with the different name and restart the ZFS-LocalPV driver to make it aware of that pool to set the corresponding node topology
+- the ZFS-LocalPV driver will look for `guid.openebs.io/14820954593456176137=true` and will remove the label from the nodes where the pool is not present
+- the ZFS-LocalPV driver will update the new node with `guid.openebs.io/14820954593456176137=true` label
+- the migrator will look for ZFSVolume resource and update the PoolName and OwnerNodeID for all the volumes.
+- the k8s scheduler will be able to see the new label and should schedule the pods to this new node.
+
+## Implementation Plan
+
+### Phase 1
+1. Implement replacement with fresh node
```

Comment:
Do we first need to support StorageClass having a zpool name pattern? prior to implementing this approach?

Reply:
no need. That is altogether different feature.

Reply:
ok. then this feature will only work by moving the pool to new node without any pools on it.

Reply:
It can have pools. Why do you ask that question, could you elaborate?